enhancements/machine-api/control-plane-machine-set.md

```diff
@@ -0,0 +1,1030 @@
+    // such as baremetal environments where additional capacity to
+    // perform rolling updates is not available.
+    Recreate ControlPlaneMachineSetStrategyType = "Recreate"
+
+    // OnDelete causes the ControlPlaneMachineSet to only replace a
+    // Machine once it has been marked for deletion. This strategy
+    // makes the rollout of updated specifications into a manual
+    // process. This allows users to test new configuration on
+    // a single Machine without forcing the rollout of all of their
+    // Control Plane Machines.
+    OnDelete ControlPlaneMachineSetStrategyType = "OnDelete"
+)
+
+// ControlPlaneMachineSetStatus represents the status of the ControlPlaneMachineSet CRD.
+type ControlPlaneMachineSetStatus struct {
+    // ObservedGeneration is the most recent generation observed for this
+    // ControlPlaneMachineSet. It corresponds to the ControlPlaneMachineSets's generation,
+    // which is updated on mutation by the API Server.
+    // +optional
+    ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+    // Replicas is the number of Control Plane Machines created by the
+    // ControlPlaneMachineSet controller.
+    // Note that during update operations this value may differ from the
+    // desired replica count.
+    Replicas int32 `json:"replicas"`
+
+    // ReadyReplicas is the number of Control Plane Machines created by the
+    // ControlPlaneMachineSet controller which are ready.
+    ReadyReplicas int32 `json:"readyReplicas,omitempty"`
+
+    // UpdatedReplicas is the number of non-terminated Control Plane Machines
+    // created by the ControlPlaneMachineSet controller that have the desired
+    // provider spec.
+    UpdatedReplicas int32 `json:"updatedReplicas,omitempty"`
+
+    // UnavailableReplicas is the number of Control Plane Machines that are
+    // still required before the ControlPlaneMachineSet reaches the desired
+    // available capacity. When this value is non-zero, the number of
+    // ReadyReplicas is less than the desired Replicas.
+    UnavailableReplicas int32 `json:"unavailableReplicas,omitempty"`
+
+    // Conditions represents the observations of the ControlPlaneMachineSet's current state.
+    // Known .status.conditions.type are: (TODO)
+    // TODO: Identify different condition types/reasons that will be needed.
+    // +patchMergeKey=type
+    // +patchStrategy=merge
+    // +listType=map
+    // +listMapKey=type
+    // +optional
+    Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+// FailureDomain represents the different configurations required to spread Machines
+// across failure domains on different platforms.
+// +union
+type FailureDomains struct {
+  // Platform identifies the platform for which the FailureDomain represents
+  // +kubebuilder:validation:Enum:="aws";"azure";"gcp";"openstack"
+  // +unionDiscriminator
+  // +optional
+  Platform configv1.PlatformType `json:platform,omitempty`
```

Comment:
API reviewer question: Because I've made this `configv1.PlatformType`, it should be one of the platform type consts, the issue here is that this list is larger than we support right now (eg include Nutanix and other platforms that won't initially be support), the other issue is that the constants of this type are camel case, so for example `AWS` instead of `aws`, `OpenStack` not `openstack`.

IIUC, the discriminator needs to match the json value of the fields in the union, so we need the lower case versions to match the json tags? Is that correct?

If it's correct, I guess I switch this back to a string and keep the kubebuilder enum validation